File: ml/publishml/recommend.py
"""
Turn a prediction into advice, by asking the model counterfactual questions.

THE NAIVE VERSION AND WHY IT IS WRONG
-------------------------------------
The obvious approach is: for each feature, nudge it, see if the score rises,
report the ones that helped. That produces garbage, for three reasons.

1. FEATURES MOVE TOGETHER. You cannot change `title_len` without changing
   `title_words`, `title_avg_word_len`, and probably `title_truncated`. Nudging
   one in isolation asks the model about a title that cannot exist, and the model
   answers confidently because nothing stopped it.

2. SOME FEATURES ARE NOT CHOICES. Video age and subscriber count predict views
   enormously well. A single-feature sweep happily advises "have more
   subscribers".

3. EXTRAPOLATION IS FREE AND MEANINGLESS. Trees are flat outside their training
   range, so pushing `title_exclaim` to 40 either does nothing or lands in a leaf
   built from three rows. Either way the number is not evidence.

WHAT THIS MODULE DOES INSTEAD
-----------------------------
Advice is generated at the level of LEVERS: a named edit a creator could actually
make, which moves a defined group of features together to a defined target. The
target is always a percentile drawn from the videos that ALREADY DID WELL in the
same niche cell, so no lever ever proposes a value outside the observed
distribution. Each lever is scored by the model's predicted change, and the ones
that clear a minimum lift are returned, largest first.

The numbers this produces are the model's opinion, not a promise, and the caller
is expected to present them that way.
"""

from __future__ import annotations

import argparse
import json
from typing import Any, Callable

import numpy as np

from . import config, features

# Minimum predicted change, in percentile points, for a lever to be worth saying
# out loud. Below ~1.5 the difference is inside the model's own noise, and a list
# of twelve suggestions each worth 0.3 points is worse than three worth 6.
MIN_LIFT = 1.5

# How many suggestions to return. More than this and the creator stops reading -
# and the tail is always the weakest advice.
MAX_SUGGESTIONS = 6


class Lever:
    """
    One editable thing, and what changing it does to the feature vector.

    `apply` receives the current feature dict and the niche target percentiles and
    returns a NEW dict. It is a function rather than a value because most edits
    are relational - "shorten the title toward 55 characters" has to recompute
    word count and truncation from the new length, not just overwrite one column.
    """

    def __init__(self, key: str, label: str, touches: list[str],
                 apply: Callable[[dict[str, float], dict[str, list[float]]], dict[str, float] | None],
                 advice: str) -> None:
        self.key = key
        self.label = label
        self.touches = touches
        self.apply = apply
        self.advice = advice


def _target(niche: dict[str, list[float]], name: str, current: float, index: int = 2) -> float | None:
    """
    The value to move a feature toward: a percentile of the successful videos.

    Returns None when the niche has no data for this feature (a cell too small to
    have a top-decile distribution), which suppresses the lever rather than
    falling back to a guess.
    """
    stats = niche.get(name)
    if not stats or index >= len(stats):
        return None
    target = float(stats[index])
    return None if abs(target - current) < 1e-9 else target


def _retitle_length(row: dict[str, float], niche: dict[str, list[float]]) -> dict[str, float] | None:
    """Move title length toward the successful median, keeping the rest coherent."""
    current = row.get("title_len", 0.0)
    target = _target(niche, "title_len", current)
    if target is None or abs(target - current) < 4:
        return None
    out = dict(row)
    out["title_len"] = target
    # Word count scales with length at roughly the current words-per-character
    # rate, so the two stay consistent instead of describing an impossible title.
    if current > 0:
        ratio = target / current
        out["title_words"] = max(1.0, round(row.get("title_words", 1.0) * ratio))
        out["title_stopword_ratio"] = row.get("title_stopword_ratio", 0.0)
    out["title_truncated"] = 1.0 if target > 60 else 0.0
    words = max(1.0, out["title_words"])
    out["title_avg_word_len"] = target / words
    return out


def _simple(name: str, index: int = 2, min_delta: float = 0.0) -> Callable[..., dict[str, float] | None]:
    """A lever that moves exactly one feature, for the genuinely independent ones."""

    def apply(row: dict[str, float], niche: dict[str, list[float]]) -> dict[str, float] | None:
        current = row.get(name, 0.0)
        target = _target(niche, name, current, index)
        if target is None or abs(target - current) <= min_delta:
            return None
        out = dict(row)
        out[name] = target
        return out

    return apply


def _thumb_text(row: dict[str, float], niche: dict[str, list[float]]) -> dict[str, float] | None:
    """Text overlay: area and block count move together, so they are one lever."""
    current = row.get("thumb_text_area", 0.0)
    target = _target(niche, "thumb_text_area", current)
    if target is None or abs(target - current) < 0.01:
        return None
    out = dict(row)
    out["thumb_text_area"] = target
    block_target = _target(niche, "thumb_text_blocks", row.get("thumb_text_blocks", 0.0))
    if block_target is not None:
        out["thumb_text_blocks"] = block_target
    return out


def _thumb_face(row: dict[str, float], niche: dict[str, list[float]]) -> dict[str, float] | None:
    """A face is present or it is not; area and count cannot disagree."""
    current_area = row.get("thumb_face_area", 0.0)
    target = _target(niche, "thumb_face_area", current_area)
    if target is None or target <= current_area + 0.01:
        return None
    out = dict(row)
    out["thumb_face_area"] = target
    out["thumb_face_count"] = max(1.0, row.get("thumb_face_count", 0.0))
    return out


def _thumb_contrast(row: dict[str, float], niche: dict[str, list[float]]) -> dict[str, float] | None:
    """Contrast, brightness spread and colourfulness are one grading decision."""
    out = dict(row)
    moved = False
    for name in ("thumb_contrast", "thumb_brightness", "thumb_saturation", "thumb_colorfulness"):
        target = _target(niche, name, row.get(name, 0.0))
        if target is not None:
            out[name] = target
            moved = True
    return out if moved else None


LEVERS = [
    Lever("title_length", "Title length", ["title_len", "title_words"], _retitle_length,
          "Rewrite the title to about {target:.0f} characters - that is the median "
          "length among videos that reached the top 10% of your niche."),
    Lever("title_number", "Number in the title", ["title_has_number"],
          _simple("title_has_number", min_delta=0.4),
          "Put a concrete number in the title (a count, a price, a year, a result)."),
    Lever("title_question", "Question framing", ["title_question"],
          _simple("title_question", min_delta=0.4),
          "Frame the title as a question the viewer wants answered."),
    Lever("title_second_person", "Speak to the viewer", ["title_second_person"],
          _simple("title_second_person", min_delta=0.4),
          "Address the viewer directly - 'you' or 'your' in the title."),
    Lever("title_curiosity", "Open a gap", ["title_curiosity"],
          _simple("title_curiosity", min_delta=0.4),
          "Add a curiosity gap: what, why, or what happened, without answering it."),
    Lever("title_caps", "Capitalisation", ["title_caps_ratio"],
          _simple("title_caps_ratio", min_delta=0.05),
          "Adjust capitalisation toward {target:.0%} of letters - shouting reads as "
          "spam in this niche, and all-lowercase reads as unfinished."),
    Lever("title_brackets", "Bracketed qualifier", ["title_brackets"],
          _simple("title_brackets", min_delta=0.4),
          "Add a bracketed qualifier - [2026], (full guide), (no code) - to carry "
          "detail without lengthening the main clause."),
    Lever("desc_first_line", "First description line", ["desc_first_line_len"],
          _simple("desc_first_line_len", min_delta=15),
          "Rewrite the first line of the description to about {target:.0f} "
          "characters. It is the only part shown before 'Show more'."),
    Lever("desc_timestamps", "Chapters", ["desc_timestamps"],
          _simple("desc_timestamps", min_delta=1),
          "Add {target:.0f} timestamped chapters. They change the surface YouTube "
          "renders and give the viewer a reason to believe the video is organised."),
    Lever("desc_length", "Description depth", ["desc_len"],
          _simple("desc_len", min_delta=150),
          "Expand the description toward {target:.0f} characters - it is what search "
          "indexes and what the top performers in your niche write."),
    Lever("tags", "Tags", ["tag_count"], _simple("tag_count", min_delta=2),
          "Use about {target:.0f} tags."),
    Lever("duration", "Length", ["duration_seconds", "duration_log"],
          _simple("duration_seconds", min_delta=60),
          "Videos in the top decile of your niche run about {target:.0f} seconds."),
    Lever("captions", "Captions", ["has_captions"], _simple("has_captions", min_delta=0.4),
          "Upload a caption track. It is indexed, and it is the single cheapest "
          "thing on this list."),
    Lever("thumb_text", "Thumbnail text", ["thumb_text_area", "thumb_text_blocks"], _thumb_text,
          "Change how much of the thumbnail is text - the successful ones in your "
          "niche sit near {target:.0%} of the frame."),
    Lever("thumb_face", "Face in the thumbnail", ["thumb_face_area", "thumb_face_count"], _thumb_face,
          "Put a face in the thumbnail at around {target:.0%} of the frame."),
    Lever("thumb_grade", "Thumbnail contrast and colour",
          ["thumb_contrast", "thumb_brightness", "thumb_saturation", "thumb_colorfulness"],
          _thumb_contrast,
          "Push contrast and saturation toward what works in your niche - the "
          "thumbnail competes at 168px wide on a phone."),
    Lever("thumb_composition", "Thumbnail composition", ["thumb_third_offset"],
          _simple("thumb_third_offset", min_delta=0.05),
          "Move the subject off dead centre, toward a rule-of-thirds intersection."),
]


def duration_coupled(row: dict[str, float]) -> dict[str, float]:
    """Keep `duration_log` and `is_shorts` consistent after a duration change."""
    import math

    seconds = row.get("duration_seconds", 0.0)
    row["duration_log"] = math.log1p(max(0.0, seconds))
    row["is_shorts"] = 1.0 if 0 < seconds <= config.SHORTS_MAX_SECONDS else 0.0
    return row


def recommend(
    predict: Callable[[dict[str, float]], float],
    row: dict[str, float],
    niche_top: dict[str, list[float]],
    max_suggestions: int = MAX_SUGGESTIONS,
    min_lift: float = MIN_LIFT,
) -> list[dict[str, Any]]:
    """
    Rank the levers by the model's predicted lift.

    `predict` takes a feature dict and returns a score, so this module never
    imports LightGBM and the same code path serves both the Python trainer's
    sanity check and any other caller.
    """
    base = predict(row)
    out: list[dict[str, Any]] = []

    for lever in LEVERS:
        changed = lever.apply(row, niche_top)
        if changed is None:
            continue
        if "duration_seconds" in lever.touches:
            changed = duration_coupled(changed)
        # Context features must be byte-identical, or the "lift" includes a change
        # to something the creator cannot do anything about.
        for name in features.CONTEXT:
            changed[name] = row.get(name, 0.0)

        lift = predict(changed) - base
        if lift < min_lift:
            continue

        primary = lever.touches[0]
        out.append(
            {
                "key": lever.key,
                "label": lever.label,
                "lift": round(lift, 2),
                "from": round(row.get(primary, 0.0), 4),
                "to": round(changed.get(primary, 0.0), 4),
                "advice": lever.advice.format(target=changed.get(primary, 0.0)),
                "touches": lever.touches,
            }
        )

    out.sort(key=lambda s: -s["lift"])
    return out[:max_suggestions]


def main(argv: list[str] | None = None) -> int:
    """Sanity-run the recommender against the exported model and dataset."""
    parser = argparse.ArgumentParser(description="Show recommendations for sample rows.")
    parser.add_argument("--rows", type=int, default=3)
    args = parser.parse_args(argv)

    model_path = config.MODEL_DIR / "publish-model.json"
    if not model_path.exists():
        raise SystemExit("No exported model. Run: python -m publishml.export")
    model = json.loads(model_path.read_text(encoding="utf-8"))
    names = model["features"]

    from .gbdt import score as gbdt_score

    def predict(row: dict[str, float]) -> float:
        return gbdt_score(model, [float(row.get(n, 0.0)) for n in names])

    dataset = config.BUILD_DIR / "dataset.jsonl"
    from .collect import read_jsonl

    shown = 0
    for entry in read_jsonl(dataset):
        if shown >= args.rows:
            break
        niche = model["nicheStats"].get(entry["cell"], {})
        top = niche.get("top") or niche.get("all") or {}
        if not top:
            continue
        suggestions = recommend(predict, entry["features"], top)
        print(f"\n{entry['id']}  cell={entry['cell']}  actual={entry['target']:.0f}  "
              f"predicted={predict(entry['features']):.1f}")
        for s in suggestions:
            print(f"  +{s['lift']:>5.2f}  {s['label']:<32} {s['advice']}")
        if not suggestions:
            print("  (nothing above the lift threshold - this row is already well set up)")
        shown += 1
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: src/lib/ml/recommend.ts
/**
 * Turn a score into advice, by asking the model counterfactual questions.
 *
 * Mirror of `ml/publishml/recommend.py`. The reasoning lives there in full; the
 * short version, because it governs every line below:
 *
 * A per-feature sweep — nudge each column, keep the ones that helped — produces
 * confident nonsense, for three reasons.
 *
 *   1. FEATURES MOVE TOGETHER. `title_len` cannot change without `title_words`,
 *      `title_avg_word_len`, and possibly `title_truncated` changing. Nudging one
 *      alone asks the model about a title that cannot exist, and it answers.
 *   2. SOME FEATURES ARE NOT CHOICES. Subscriber count and video age predict views
 *      superbly, so a naive sweep advises "have more subscribers".
 *   3. EXTRAPOLATION IS FREE. Trees are flat outside their training range, so
 *      pushing `title_exclaim` to 40 lands in a leaf built from three rows.
 *
 * So advice is generated at the level of a LEVER: a named edit a creator could
 * actually make, moving a coherent group of features to a target drawn from the
 * videos that ALREADY reached the top decile of the same niche cell. No lever can
 * propose a value nobody in that niche has achieved, and no lever can touch a
 * CONTEXT feature — those are re-pinned to the original row before the model is
 * asked, so the reported lift is only ever attributable to the edit itself.
 */

import { CONTEXT_FEATURES, SHORTS_MAX_SECONDS, type FeatureRow } from './features';
import { scoreNamed } from './gbdt';
import { modelState, type NicheDistribution } from './index';

/**
 * Minimum predicted change, in percentile points, worth saying out loud.
 *
 * Below roughly this the difference sits inside the model's own noise, and twelve
 * suggestions worth 0.3 points each are worse than three worth 6 — they cost the
 * creator the same reading time and carry no information.
 */
export const MIN_LIFT = 1.5;

/** Past this the creator stops reading, and the tail is always the weakest advice. */
export const MAX_SUGGESTIONS = 6;

/** Index into `PublishModel.percentiles` (`[10, 25, 50, 75, 90]`) meaning p50. */
const P50 = 2;

export type Suggestion = {
  key: string;
  label: string;
  /** Predicted change in the 0–100 niche percentile. The model's opinion, not a promise. */
  lift: number;
  from: number;
  to: number;
  advice: string;
  /** Every feature this edit moves — so a UI can show what it actually changed. */
  touches: string[];
};

type Applier = (row: FeatureRow, niche: NicheDistribution) => FeatureRow | null;

type Lever = {
  key: string;
  label: string;
  touches: string[];
  apply: Applier;
  /** `{target}` is substituted with the new value of `touches[0]`. */
  advice: string;
  /** How to render `{target}`: a rounded count, or a percentage of the frame. */
  format: 'count' | 'percent';
};

/**
 * The value to move a feature toward: a percentile of the videos that succeeded.
 *
 * `null` when the niche has no distribution for this feature — a cell too thin to
 * have a top-decile sample. That suppresses the lever entirely rather than falling
 * back to a guess, which is why a creator in an unusual niche gets fewer
 * suggestions rather than worse ones.
 */
function target(
  niche: NicheDistribution,
  name: string,
  current: number,
  index: number = P50,
): number | null {
  const stats = niche[name];
  if (!stats || index >= stats.length) return null;
  const value = stats[index];
  if (!Number.isFinite(value)) return null;
  return Math.abs(value - current) < 1e-9 ? null : value;
}

/** Move title length toward the successful median, keeping the rest coherent. */
function retitleLength(row: FeatureRow, niche: NicheDistribution): FeatureRow | null {
  const current = row.title_len ?? 0;
  const goal = target(niche, 'title_len', current);
  // Under four characters is not a rewrite anyone would make, and the model
  // cannot resolve a difference that small anyway.
  if (goal === null || Math.abs(goal - current) < 4) return null;

  const out: FeatureRow = { ...row };
  out.title_len = goal;
  if (current > 0) {
    // Word count scales with length at the title's own current words-per-character
    // rate, so the counterfactual describes a title that could exist.
    const ratio = goal / current;
    out.title_words = Math.max(1, Math.round((row.title_words ?? 1) * ratio));
  }
  out.title_truncated = goal > 60 ? 1 : 0;
  out.title_avg_word_len = goal / Math.max(1, out.title_words);
  return out;
}

/** A lever that moves exactly one feature — only for the genuinely independent ones. */
function simple(name: string, minDelta = 0, index: number = P50): Applier {
  return (row, niche) => {
    const current = row[name] ?? 0;
    const goal = target(niche, name, current, index);
    if (goal === null || Math.abs(goal - current) <= minDelta) return null;
    return { ...row, [name]: goal };
  };
}

/** Text overlay: area and block count are one design decision, not two. */
function thumbText(row: FeatureRow, niche: NicheDistribution): FeatureRow | null {
  const current = row.thumb_text_area ?? 0;
  const goal = target(niche, 'thumb_text_area', current);
  if (goal === null || Math.abs(goal - current) < 0.01) return null;
  const out: FeatureRow = { ...row, thumb_text_area: goal };
  const blocks = target(niche, 'thumb_text_blocks', row.thumb_text_blocks ?? 0);
  if (blocks !== null) out.thumb_text_blocks = blocks;
  return out;
}

/** A face is present or it is not; area and count cannot disagree. */
function thumbFace(row: FeatureRow, niche: NicheDistribution): FeatureRow | null {
  const currentArea = row.thumb_face_area ?? 0;
  const goal = target(niche, 'thumb_face_area', currentArea);
  // Only ever suggests adding or enlarging a face. "Make the face smaller" is
  // advice the geometric features are far too crude to support.
  if (goal === null || goal <= currentArea + 0.01) return null;
  return {
    ...row,
    thumb_face_area: goal,
    thumb_face_count: Math.max(1, row.thumb_face_count ?? 0),
  };
}

/** Contrast, brightness, saturation and colourfulness are one grading decision. */
function thumbGrade(row: FeatureRow, niche: NicheDistribution): FeatureRow | null {
  const out: FeatureRow = { ...row };
  let moved = false;
  for (const name of [
    'thumb_contrast',
    'thumb_brightness',
    'thumb_saturation',
    'thumb_colorfulness',
  ]) {
    const goal = target(niche, name, row[name] ?? 0);
    if (goal !== null) {
      out[name] = goal;
      moved = true;
    }
  }
  return moved ? out : null;
}

/**
 * Every lever, in declaration order. Ranking is by predicted lift, so this order
 * only decides ties.
 */
export const LEVERS: readonly Lever[] = [
  {
    key: 'title_length',
    label: 'Title length',
    touches: ['title_len', 'title_words', 'title_avg_word_len', 'title_truncated'],
    apply: retitleLength,
    advice:
      'Rewrite the title to about {target} characters — the median length among videos that reached the top 10% of your niche.',
    format: 'count',
  },
  {
    key: 'title_number',
    label: 'Number in the title',
    touches: ['title_has_number'],
    apply: simple('title_has_number', 0.4),
    advice: 'Put a concrete number in the title — a count, a price, a year, a result.',
    format: 'count',
  },
  {
    key: 'title_question',
    label: 'Question framing',
    touches: ['title_question'],
    apply: simple('title_question', 0.4),
    advice: 'Frame the title as a question the viewer wants answered.',
    format: 'count',
  },
  {
    key: 'title_second_person',
    label: 'Speak to the viewer',
    touches: ['title_second_person'],
    apply: simple('title_second_person', 0.4),
    advice: "Address the viewer directly — 'you' or 'your' in the title.",
    format: 'count',
  },
  {
    key: 'title_curiosity',
    label: 'Open a gap',
    touches: ['title_curiosity'],
    apply: simple('title_curiosity', 0.4),
    advice: 'Add a curiosity gap: what, why, or what happened — without answering it.',
    format: 'count',
  },
  {
    key: 'title_caps',
    label: 'Capitalisation',
    touches: ['title_caps_ratio'],
    apply: simple('title_caps_ratio', 0.05),
    advice:
      'Adjust capitalisation toward {target} of letters — shouting reads as spam in this niche, and all-lowercase reads as unfinished.',
    format: 'percent',
  },
  {
    key: 'title_brackets',
    label: 'Bracketed qualifier',
    touches: ['title_brackets'],
    apply: simple('title_brackets', 0.4),
    advice:
      'Add a bracketed qualifier — [2026], (full guide), (no code) — to carry detail without lengthening the main clause.',
    format: 'count',
  },
  {
    key: 'desc_first_line',
    label: 'First description line',
    touches: ['desc_first_line_len'],
    apply: simple('desc_first_line_len', 15),
    advice:
      "Rewrite the first line of the description to about {target} characters. It is the only part shown before 'Show more'.",
    format: 'count',
  },
  {
    key: 'desc_timestamps',
    label: 'Chapters',
    touches: ['desc_timestamps'],
    apply: simple('desc_timestamps', 1),
    advice:
      'Add {target} timestamped chapters. They change the surface YouTube renders and give the viewer a reason to believe the video is organised.',
    format: 'count',
  },
  {
    key: 'desc_length',
    label: 'Description depth',
    touches: ['desc_len'],
    apply: simple('desc_len', 150),
    advice:
      'Expand the description toward {target} characters — it is what search indexes, and what the top performers in your niche write.',
    format: 'count',
  },
  {
    key: 'tags',
    label: 'Tags',
    touches: ['tag_count'],
    apply: simple('tag_count', 2),
    advice: 'Use about {target} tags.',
    format: 'count',
  },
  {
    key: 'duration',
    label: 'Length',
    touches: ['duration_seconds', 'duration_log', 'is_shorts'],
    apply: simple('duration_seconds', 60),
    advice: 'Videos in the top decile of your niche run about {target} seconds.',
    format: 'count',
  },
  {
    key: 'captions',
    label: 'Captions',
    touches: ['has_captions'],
    apply: simple('has_captions', 0.4),
    advice:
      'Upload a caption track. It is indexed, and it is the single cheapest thing on this list.',
    format: 'count',
  },
  {
    key: 'thumb_text',
    label: 'Thumbnail text',
    touches: ['thumb_text_area', 'thumb_text_blocks'],
    apply: thumbText,
    advice:
      'Change how much of the thumbnail is text — the successful ones in your niche sit near {target} of the frame.',
    format: 'percent',
  },
  {
    key: 'thumb_face',
    label: 'Face in the thumbnail',
    touches: ['thumb_face_area', 'thumb_face_count'],
    apply: thumbFace,
    advice: 'Put a face in the thumbnail at around {target} of the frame.',
    format: 'percent',
  },
  {
    key: 'thumb_grade',
    label: 'Thumbnail contrast and colour',
    touches: ['thumb_contrast', 'thumb_brightness', 'thumb_saturation', 'thumb_colorfulness'],
    apply: thumbGrade,
    advice:
      'Push contrast and saturation toward what works in your niche — the thumbnail competes at 168px wide on a phone.',
    format: 'percent',
  },
  {
    key: 'thumb_composition',
    label: 'Thumbnail composition',
    touches: ['thumb_third_offset'],
    apply: simple('thumb_third_offset', 0.05),
    advice: 'Move the subject off dead centre, toward a rule-of-thirds intersection.',
    format: 'count',
  },
];

/**
 * Keep `duration_log` and `is_shorts` consistent after a duration change.
 *
 * Without this a "make it 8 minutes" counterfactual leaves `is_shorts = 1` and a
 * `duration_log` from the old value, and the model is asked about an eight-minute
 * Short — a thing that does not exist and whose leaves are therefore arbitrary.
 */
export function durationCoupled(row: FeatureRow): FeatureRow {
  const seconds = row.duration_seconds ?? 0;
  return {
    ...row,
    duration_log: Math.log1p(Math.max(0, seconds)),
    is_shorts: seconds > 0 && seconds <= SHORTS_MAX_SECONDS ? 1 : 0,
  };
}

/** Substitute `{target}` in a lever's advice string. */
function fillAdvice(advice: string, value: number, format: Lever['format']): string {
  const rendered =
    format === 'percent' ? `${Math.round(value * 100)}%` : `${Math.round(value)}`;
  return advice.replace('{target}', rendered);
}

export type RecommendOptions = {
  maxSuggestions?: number;
  minLift?: number;
};

/**
 * Rank the levers by the model's predicted lift, largest first.
 *
 * `predict` is injected rather than imported so this function can be tested
 * against a known scoring function, and so the same code serves the exported
 * model, a fixture, and anything else that can score a row.
 */
export function recommend(
  predict: (row: FeatureRow) => number,
  row: FeatureRow,
  nicheTop: NicheDistribution,
  options: RecommendOptions = {},
): Suggestion[] {
  const maxSuggestions = options.maxSuggestions ?? MAX_SUGGESTIONS;
  const minLift = options.minLift ?? MIN_LIFT;
  const base = predict(row);
  const out: Suggestion[] = [];

  for (const lever of LEVERS) {
    let changed = lever.apply(row, nicheTop);
    if (changed === null) continue;
    if (lever.touches.includes('duration_seconds')) changed = durationCoupled(changed);

    // Re-pin every context feature to the original value. An applier has no reason
    // to touch one, but the guarantee has to hold by construction rather than by
    // review — otherwise a future lever could quietly start reporting the lift from
    // a subscriber count the creator cannot change.
    for (const name of CONTEXT_FEATURES) changed[name] = row[name] ?? 0;

    const lift = predict(changed) - base;
    if (lift < minLift) continue;

    const primary = lever.touches[0];
    const to = changed[primary] ?? 0;
    out.push({
      key: lever.key,
      label: lever.label,
      lift: Math.round(lift * 100) / 100,
      from: Math.round((row[primary] ?? 0) * 10_000) / 10_000,
      to: Math.round(to * 10_000) / 10_000,
      advice: fillAdvice(lever.advice, to, lever.format),
      touches: lever.touches,
    });
  }

  out.sort((a, b) => b.lift - a.lift);
  return out.slice(0, maxSuggestions);
}

export type NicheAdvice = {
  suggestions: Suggestion[];
  /** The cell the advice was drawn from, and how many videos stood behind it. */
  cell: string;
  cellSize: number;
  topSize: number;
  /**
   * How many levers had a target in this niche and produced a coherent edit.
   *
   * Zero and non-zero mean completely different things when `suggestions` is empty:
   * zero is "this niche has no distribution to aim at", non-zero is "we tried N edits
   * and the model would not back any of them".
   */
  considered: number;
  /**
   * The largest lift that fell below the threshold, or null if none did.
   *
   * Carried so a caller can say "no single change is predicted to move the score by
   * more than 0.6 points" instead of rendering an unexplained blank. Suppressing the
   * weak advice is right; suppressing the fact that it was weak is not.
   */
  bestRejectedLift: number | null;
};

/**
 * Advice for a row in a named niche cell, using the loaded model.
 *
 * `null` — not an empty list — when there is no model, no such cell, or the cell
 * has no top-decile distribution. An empty list means "we looked and found nothing
 * worth changing", which is a completely different message to show a creator.
 */
export function adviseInNiche(
  row: FeatureRow,
  cell: string,
  options: RecommendOptions = {},
): NicheAdvice | null {
  const state = modelState();
  if (!state.available) return null;
  const niche = state.model.nicheStats[cell];
  if (!niche) return null;
  // `top` is empty for cells with fewer than ten top-decile videos. Falling back to
  // `all` would mean advising the creator toward the niche's average video, which
  // is the opposite of the point.
  const distribution = niche.top && Object.keys(niche.top).length > 0 ? niche.top : null;
  if (!distribution) return null;

  // Raw ensemble output, deliberately NOT clamped to 0-100. A lift is a
  // difference between two predictions, and clamping both ends would silently
  // report 0 for a real improvement on a row already predicted near 100.
  const model = state.model;
  const predict = (candidate: FeatureRow): number => scoreNamed(model, candidate);

  // One pass over every lever, unfiltered, then split. Running `recommend` twice with
  // different thresholds would double the tree walks for no reason, and filtering here
  // is what makes `considered` and `bestRejectedLift` free.
  const everything = recommend(predict, row, distribution, {
    minLift: -Infinity,
    maxSuggestions: Number.MAX_SAFE_INTEGER,
  });
  const minLift = options.minLift ?? MIN_LIFT;
  const maxSuggestions = options.maxSuggestions ?? MAX_SUGGESTIONS;
  const rejected = everything.filter((s) => s.lift < minLift);

  return {
    suggestions: everything.filter((s) => s.lift >= minLift).slice(0, maxSuggestions),
    cell,
    cellSize: niche.n,
    topSize: niche.nTop,
    considered: everything.length,
    // `everything` is already sorted by lift descending, so the first rejection is the
    // best one.
    bestRejectedLift: rejected.length > 0 ? rejected[0].lift : null,
  };
}
